temporal/fx.go

```diff
@@ -373,31 +379,76 @@ func (params ServiceProviderParamsCommon) getCommonServiceOptions(serviceName pr
 			params.SpanExporters,
 		),
 		fx.Provide(
-			func() persistenceClient.AbstractDataStoreFactory { return params.DataStoreFactory },
-			func() client.FactoryProvider { return params.ClientFactoryProvider },
-			func() authorization.JWTAudienceMapper { return params.AudienceGetter },
-			func() resolver.ServiceResolver { return params.PersistenceServiceResolver },
-			func() searchattribute.Mapper { return params.SearchAttributesMapper },
-			func() []grpc.UnaryServerInterceptor { return params.CustomInterceptors },
-			func() authorization.Authorizer { return params.Authorizer },
-			func() authorization.ClaimMapper { return params.ClaimMapper },
-			func() encryption.TLSConfigProvider { return params.TlsConfigProvider },
-			func() dynamicconfig.Client { return params.DynamicConfigClient },
-			func() log.Logger { return params.Logger },
 			resource.DefaultSnTaggedLoggerProvider,
+			params.PersistenceFactoryProvider,
+			func() persistenceClient.AbstractDataStoreFactory {
+				return params.DataStoreFactory
+			},
+			func() client.FactoryProvider {
+				return params.ClientFactoryProvider
+			},
+			func() authorization.JWTAudienceMapper {
+				return params.AudienceGetter
+			},
+			func() resolver.ServiceResolver {
+				return params.PersistenceServiceResolver
+			},
+			func() searchattribute.Mapper {
+				return params.SearchAttributesMapper
+			},
+			func() []grpc.UnaryServerInterceptor {
+				return params.CustomInterceptors
+			},
+			func() authorization.Authorizer {
+				return params.Authorizer
+			},
+			func() authorization.ClaimMapper {
+				return params.ClaimMapper
+			},
+			func() encryption.TLSConfigProvider {
+				return params.TlsConfigProvider
+			},
+			func() dynamicconfig.Client {
+				return params.DynamicConfigClient
+			},
+			func() log.Logger {
+				return params.Logger
+			},
 			func() metrics.Handler {
 				return params.MetricsHandler.WithTags(metrics.ServiceNameTag(serviceName))
 			},
-			func() esclient.Client { return params.EsClient },
-			func() resource.NamespaceLogger { return params.NamespaceLogger },
-			params.PersistenceFactoryProvider,
+			func() esclient.Client {
+				return params.EsClient
+			},
+			func() resource.NamespaceLogger {
+				return params.NamespaceLogger
+			},
+			func() tasks.TaskCategoryRegistry {
+				return params.TaskCategoryRegistry
+			},
 		),
 		ServiceTracingModule,
 		resource.DefaultOptions,
 		FxLogAdapter,
 	)
 }
 
+// TaskCategoryRegistryProvider provides an immutable tasks.TaskCategoryRegistry to the server, which is intended to be
+// shared by each service. Why do we need to initialize this at the top-level? Because, even though the presence of the
+// archival task category is only needed by the history service, which must conditionally start a queue processor for
+// it, we also do validation on request task categories in the frontend service. As a result, we need to initialize the
```

Comment:
Why task category validation needs to be done in frontend? I thought all task related requests will be forwarded to history and validation can be done there. 

Maybe I missed some cases? 
ListHistoryTasks doesn't count. That one should be moved/forwarded to history. Frontend should not access history service persistence directly.

Reply:
It's only `ListHistoryTasks`. The `GetCategories` method isn't called anywhere outside of history, fortunately. However, the usage of `GetCategoryByID` in `ListHistoryTasks` is used for more than just validation; we also need to fetch the `Category` object itself because, down the stack in `ListHistoryTasks`, we use the other properties of the `Category`:
1. [Usage of `Category.ID()`](https://github.com/temporalio/temporal/blob/a3ae60298bb165f60b7181a5944cf4cfe089068e/common/persistence/cassandra/mutable_state_task_store.go#L784)
2. [Usage of `Category.Name()`](https://github.com/temporalio/cds/blob/1ad145152b5c89b7f76f2df2b6e91128413d8ded/persistence/metrics/metrics_managers.go#L53)

So, until this is fixed where the frontend service forwards to the history service, we still need the task category registry in the frontend service.

Reply:
hmmm, ok. Looks like we have to move ListHistoryTasks to history service first before removing the registry from frontend. Would you mind create a task for ^?

Reply:
Done